Ark.Tools.Sql/AbstractSqlContextAsync.cs

```diff
@@ -0,0 +1,110 @@
+using System.Data;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Ark.Tools.Sql
+{
+    public abstract class AbstractSqlContextAsync<Tag> : ISqlContextAsync
+    {
+        private DbConnection? _connection;
+        private DbTransaction? _transaction;
+        private IsolationLevel _isolationLevel;
+
+        public AbstractSqlContextAsync(DbConnection connection, DbTransaction transaction, IsolationLevel isolationLevel = IsolationLevel.ReadCommitted)
+        {
+            _connection = connection ?? throw new ArgumentNullException(nameof(connection));
+            _isolationLevel = isolationLevel;
+            _transaction = transaction;
+        }
+
+        public AbstractSqlContextAsync(DbConnection connection, IsolationLevel isolationLevel = IsolationLevel.ReadCommitted)
+        {
+            _connection = connection ?? throw new ArgumentNullException(nameof(connection));
+            _isolationLevel = isolationLevel;
+            _transaction = _connection.BeginTransaction(isolationLevel);
+        }
+
+        public AbstractSqlContextAsync(DbTransaction transaction)
+        {
+            _transaction = transaction ?? throw new ArgumentNullException(nameof(transaction));
+            _connection = transaction.Connection ?? throw new ArgumentNullException(nameof(transaction.Connection));
+            _isolationLevel = transaction.IsolationLevel;
+        }
+
+
+        public DbConnection Connection
+        {
+            get
+            {
+                return _connection ?? throw new InvalidProgramException("There's no connection");
+            }
+        }
+
+        public DbTransaction Transaction
+        {
+            get
+            {
+                return _transaction ?? throw new InvalidProgramException("There's no transaction");
+            }
+        }
+
+        public virtual async ValueTask CommitAysnc(CancellationToken ctk)
+        {
+#if !(NETSTANDARD2_0 || NET472)
+            if (_transaction != null)
+            {
+                await _transaction.CommitAsync(ctk);
+                await _transaction.DisposeAsync();
+            }
+            _transaction = null;
+#else
+            throw new NotSupportedException("Async SQL not supported for this .NET framework");
+#endif
+        }
+
+        public async ValueTask ChangeIsolationLevelAsync(IsolationLevel isolationLevel, CancellationToken ctk)
+        {
+#if !(NETSTANDARD2_0 || NET472)
+            _isolationLevel = isolationLevel;
+            await RollbackAsync(ctk);
+#else
+            throw new NotSupportedException("Async SQL not supported for this .NET framework");
+#endif
+        }
+
+        public async ValueTask RollbackAsync(CancellationToken ctk)
+        {
+#if !(NETSTANDARD2_0 || NET472)
+            if (_transaction != null)
+            {
+                await _transaction.RollbackAsync(ctk);
+                await _transaction.DisposeAsync();
+            }
+
+            _transaction = null;
```

Comment:
and what happens on the next 'query' ?

please test the following Handler with the TestContext

```cs
await using ctx = await _factory.CreateAsync(ctk);
await ctx.ChangeIsolationLevel(Serializable);
await ctx.ReadSomething(); // this now throws, you cannot really change the IsolationLevel

```